File: Perl/extractSNPs.pl
#!/usr/bin/perl
use strict;
use warnings;
use List::Util qw( first );

# change here for different trait and different NAMs
my $trait = "TotalSN";
my $nam = "DD";

my @raw = glob("pvalues*$nam*");
my %pvalues;

open(OUT,">${trait}_${nam}.txt") or die "Cannot write to the file!\n";
print OUT "SNP";

foreach my $file (@raw){
	print OUT "\t$file";
	open (RAW, $file);
	chomp(my $header = <RAW>); # the first line
	my @ID = split(/\t/, $header);
	my $n = first {$ID[$_] eq $trait} 0 .. $#ID; # the column # of the trait
	while (<RAW>) {
		chomp(my $line = $_);
		my @col = split(/\t/, $line);
		my $key = $col[0];
		my $value = $col[$n];
		next if $value > 0.01;
		$pvalues{$key} .= "\t$value";
	}
}

print OUT "\n";


while ( my($key, $value) = each %pvalues ){
  my $count = () = $value =~ /\t/g; # count tab occurences
  if ($count > 1) {print OUT "$key$value\n";} # only SNPs significant in more than 1 environment were printed out
}

close OUT;

File: Python/extractSNPs.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#  Copyright 2016 Junli Zhang <zhjl86@gmail.com>
#  
from glob import glob

# change here for different trait and different NAMs
trait = "TotalSN"
nam = "DD"

raw = glob("pvalues*" + nam + "*")
raw.sort()

out = open(trait + "_" + nam + ".txt", "w")
out.write("SNP") # the first line of output
pvalues = {}

for ff in raw:
	out.write("\t" + ff)
	with open(ff) as infile:
		#next(infile) # skip header
		ID = infile.readline().rstrip() # the first line is the trait names
		n1 = ID.split('\t').index(trait) # the column of the trait
		for line in infile:
			col = line.rstrip().split('\t')
			#key = ",".join(col[0:3])
			key = col[0]
			if float(col[n1]) > 0.01: continue # skip the line if p> 0.01
			pvalues.setdefault(key, "") # avoid key error for the first time
			pvalues[key] += '\t' + col[n1]

out.write("\n")
for k, v in pvalues.iteritems():
	if v.count('\t') > 1:
		out.write(k + v + "\n")
out.close()
